### ISSUE TITLE
autoclonetype2 breaks GenericParameterizedBundle 

### ISSUE THREAD

**Issue by sequencer** [chipsalliance/chisel3#1858]
**Type of issue**: bug report

**Impact**: no functional change

**Development Phase**: request

**If the current behavior is a bug, please provide the steps to reproduce the problem:**
See https://github.com/sequencer/autoclonetype2_bug

**Please tell us about your environment:**
depends on: Arch Linux `riscv64-elf-gcc` `riscv64-elf-newlib`
if not on Arch Linux: modify this [line](https://github.com/sequencer/autoclonetype2_bug/blob/master/playground/src/CodesignConfig.scala#L48)

**Comment by jackkoenig** [comment]
Having looked into this, autoclonetype2 doesn't break [GenericParameterizedBundle](https://github.com/chipsalliance/rocket-chip/blob/025184e1a0fdd8338eb0f66ad9ab33587b619854/src/main/scala/util/GenericParameterizedBundle.scala#L7), it does sort of break [XilinxVCU118MIGPads](https://github.com/sifive/fpga-shells/blob/11055ec7cb73dbb4d1b67cd583c2870ba27c71b7/src/main/scala/devices/xilinx/xilinxvcu118mig/XilinxVCU118MIG.scala#L18)'s *use* of `GenericParameterizedBundle`, but it's currently only correct by coicidence and there's nothing reasonable that autoclonetype2 could do to not break this. `XilinxVCU118MIGPads` has 2 constructors while `GenericParameterizedBundle` just reflectively picks the first one it can find. This is relying on the order of the constructors remaining consistent which it apparently is not safe to rely on. autoclonetype2 doesn't do anything to change this order (it doesn't like, reorder them by accident), all it does is add 2 additional methods which apparently is enough to change the order of the constructors returned by runtime reflection.

This can be solved in a couple of ways:
1. Make `GenericParameterizedBundle` smarter (actually pick the right constructor or error if there are more than 1)
2. Stop using `GenericParameterizedBundle` (I intend to delete it's implementation of `cloneType` in rocket-chip soon)

TLDR, not a bug in autoclonetype2.
